lib/shape.ml

```diff
@@ -29,8 +32,36 @@ let rectangle ?x ?y length width =
 
 let ellipse ?x ?y rx ry =
   match (x, y) with
-  | Some x, Some y -> Ellipse {c = {x; y}; rx; ry}
-  | _ -> Ellipse {c = { x = canvas_mid.x; y = canvas_mid.y}; rx; ry}
+  | Some x, Some y -> Ellipse { c = { x; y }; rx; ry }
+  | _ -> Ellipse { c = { x = canvas_mid.x; y = canvas_mid.y }; rx; ry }
+
+let triangle ?x1 ?y1 ?x2 ?y2 ?x3 ?y3 () =
+  let p1 =
+    match (x1, y1) with
+    | Some x, Some y -> { x; y }
+    | _ -> canvas_mid
+  in
+  let p2 =
+    match (x2, y2) with
+    | Some x, Some y -> { x; y }
+    | _ -> canvas_mid
+  in
+  let p3 =
+    match (x3, y3) with
+    | Some x, Some y -> { x; y }
+    | _ -> canvas_mid
+  in
+  Triangle { p1; p2; p3 }
+
+let translate dx dy shape =
```

Comment:
The implementation looks fine to me. You'll need to add this to function to the `.mli` file.

Reply:
I'm not sure how to add this. From my understanding above, I thought you wanted me to completely take out the implementation of the triangle translate transformation function. 

Or do I leave the implementation and only edit the addition in the shape.mli file?
If so, how is it supposed to look like?

Reply:
Yes, I meant add `val translate: int -> int -> shape -> shape` to the `.mli` file.